base/random.jl

```diff
@@ -683,6 +683,44 @@ end
 rand{T}(rng::AbstractRNG, r::AbstractArray{T}, dims::Dims) = rand!(rng, Array{T}(dims), r)
 rand(rng::AbstractRNG, r::AbstractArray, dims::Int...) = rand(rng, r, dims)
 
+# rand from a string
+
+rand(rng::AbstractRNG, s::AbstractString) = rand_iter(rng, s, Base.iteratorsize(s))
+rand(s::AbstractString) = rand(GLOBAL_RNG, s)
+
+## optimization for String
+
+isvalid_unsafe(s::String, i) = !Base.is_valid_continuation(unsafe_load(pointer(s), i))
+
+function rand(rng::AbstractRNG, s::String)
+    rg = RangeGenerator(1:s.len)
+    while true
+        pos = rand(rng, rg)
+        isvalid_unsafe(s, pos) && return s[pos]
+    end
+end
+
+## rand from an iterator
+
+rand_iter(rng, s, ::Base.IteratorSize) = throw(ArgumentError("iterator must have a known length"))
+
+function rand_iter(rng::AbstractRNG, s, ::Union{Base.HasShape,Base.HasLength})::eltype(s)
+    pos = rand(rng, 1:length(s))
+    for (i, c) in enumerate(s)
+        i == pos && return c
+    end
+end
```

Comment:
I'm a little skeptical of the `rand_iter` function and `AbstractString` support.    People using `rand` with other string types will probably not expect this function to be O(n); it seems preferable to throw a MethodError rather than to use this implementation.